src/main/flight/pid.c

```diff
@@ -207,55 +210,67 @@ void pidInitFilters(const pidProfile_t *pidProfile)
 
     if (dTermNotchHz != 0 && pidProfile->dterm_notch_cutoff != 0) {
         static biquadFilter_t biquadFilterNotch[2];
-        dtermNotchFilterApplyFn = (filterApplyFnPtr)biquadFilterApply;
+        dtermNotchApplyFn = (filterApplyFnPtr)biquadFilterApply;
         const float notchQ = filterGetNotchQ(dTermNotchHz, pidProfile->dterm_notch_cutoff);
         for (int axis = FD_ROLL; axis <= FD_PITCH; axis++) {
-            dtermFilterNotch[axis] = &biquadFilterNotch[axis];
-            biquadFilterInit(dtermFilterNotch[axis], dTermNotchHz, targetPidLooptime, notchQ, FILTER_NOTCH);
+            dtermNotch[axis] = &biquadFilterNotch[axis];
+            biquadFilterInit(dtermNotch[axis], dTermNotchHz, targetPidLooptime, notchQ, FILTER_NOTCH);
         }
     } else {
-        dtermNotchFilterApplyFn = nullFilterApply;
+        dtermNotchApplyFn = nullFilterApply;
     }
-
-    static dtermFilterLpf_t dtermFilterLpfUnion;
-    if (pidProfile->dterm_lpf_hz == 0 || pidProfile->dterm_lpf_hz > pidFrequencyNyquist) {
-        dtermLpfApplyFn = nullFilterApply;
+    
+    //2nd Dterm Lowpass Filter
+	static pt1Filter_t dtermLowpass2State;
+	if (pidProfile->dterm_lowpass2_hz == 0 || pidProfile->dterm_lowpass2_hz > pidFrequencyNyquist) {
+		dtermLowpass2ApplyFn = nullFilterApply;
+	} else {
+		dtermLowpass2ApplyFn = (filterApplyFnPtr)pt1FilterApply;
+		for (int axis = FD_ROLL; axis <= FD_PITCH; axis++) {
+			dtermLowpass2[axis] = &dtermLowpass2State;
+			pt1FilterInit(dtermLowpass2[axis], pidProfile->dterm_lowpass2_hz, dT);
+		}
+	}
+
+    static dtermLowpass_t dtermLowpassUnion;
```

Comment:
Same here, this effectively makes state shared for all axes

Reply:
OK that was a mistake, unintended, need to fix that, sorry

Reply:
I checked, I've just used exactly the same method as for the yaw code and the other filter code as in release 3.3.   Hope there's nothing fundamentally wrong, but if there is, we will need to fix it.

Reply:
Yaw is just one axis, so it's not shared by definition.
In your code roll&pitch get shared Dterm state for lowpass.

Reply:
Lol at how it's done in Fujin's PR for gyro lowpass, you need to create state for PID filters in advance. I can help when I get home

Reply:
OK thanks I'll need your help with that.  I can see something he did in the gyro lowpass code, calculating pt1FilterGain, but looks like I'd have to implement a lot of his PR just to make this work!  Hopefully it's not going to be *that* difficult!  In the meantime, it seems to fly OK - out of curiosity, what would be the effect of this error?

Reply:
Coupling of roll and pitch, spikes and dips in roll will have effect on pitch at the same sample, and the same in pitch will have an effect on roll at the next sample

Reply:
Maybe the simplest, but ugliest, solution would be to run the code in pid.c on a per axis basis.  That would work.  Later after @fujin's PR he could use his gyro code to tidy up this PR?

Reply:
@DieHertz: https://github.com/betaflight/betaflight/pull/5427/files#diff-c5f9576172e273033324ba218524cc97R181

dtermLowpass_t contains arrays ...

Reply:
Okay then. I don't like this approach though.

Reply:
So it is OK?  I don't need to do anything now about that static issue?

Reply:
Yes, apologies

Reply:
3.3 is ok, your new code is not

Reply:
oh

Reply:
is it obvious what I need to do to fix it?

Reply:
Yes, add another static variable for lowpass2 state and use it for initialization of lowpass2

Reply:
@ctzsnooze : the problem is here: https://github.com/betaflight/betaflight/pull/5427/files#r173423911

Reply:
OK sorry I have to wait for actual code suggestion, too dumb to know how to do that.

Reply:
Would this work:

    //2nd Dterm Lowpass Filter      
    if (pidProfile->dterm_lowpass2_hz == 0 || pidProfile->dterm_lowpass2_hz > pidFrequencyNyquist) {
    	dtermLowpass2ApplyFn = nullFilterApply;
    } else {
    	static pt1Filter_t dtermLowpass2State[2];  
        dtermLowpass2ApplyFn = (filterApplyFnPtr)pt1FilterApply;
        for (int axis = FD_ROLL; axis <= FD_PITCH; axis++) {
            dtermLowpass2[axis] = &dtermLowpass2State[axis];
            pt1FilterInit(dtermLowpass2[axis], pidProfile->dterm_lowpass2_hz, dT);
        }        
    }

Reply:
dtermLowpass_t contains 2-element arrays for filter state (roll+pitch)
pt1Filter_t is single state

Use `static pt1Filter_t dtermLowpass2State[2];` or define type

Reply:
I just pushed this, hope it is now OK?

    //2nd Dterm Lowpass Filter   
    static pt1Filter_t dtermFilterLowpass2[2];     
    if (pidProfile->dterm_lowpass2_hz == 0 || pidProfile->dterm_lowpass2_hz > pidFrequencyNyquist) {
    	dtermLowpass2ApplyFn = nullFilterApply;
    } else {
        dtermLowpass2ApplyFn = (filterApplyFnPtr)pt1FilterApply;
        for (int axis = FD_ROLL; axis <= FD_PITCH; axis++) {
            dtermLowpass2[axis] = &dtermFilterLowpass2[axis];
            pt1FilterInit(dtermLowpass2[axis], pidProfile->dterm_lowpass2_hz, dT);
        }        
    }